sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/NativeMethodsShouldBeWrapped.cs

```diff
@@ -45,33 +45,31 @@ public sealed class NativeMethodsShouldBeWrapped : SonarDiagnosticAnalyzer
 
         protected override void Initialize(SonarAnalysisContext context)
         {
-            context.RegisterSyntaxNodeAction(ReportPublicExternalMethods, SyntaxKind.MethodDeclaration);
+            context.RegisterSymbolAction(ReportPublicExternalMethods, SymbolKind.Method);
             context.RegisterSyntaxNodeAction(ReportTrivialWrappers, SyntaxKind.MethodDeclaration);
         }
 
-        private static void ReportPublicExternalMethods(SyntaxNodeAnalysisContext c)
+        private static void ReportPublicExternalMethods(SymbolAnalysisContext c)
         {
-            var methodDeclaration = (MethodDeclarationSyntax)c.Node;
-            var methodSymbol = c.SemanticModel.GetDeclaredSymbol(methodDeclaration);
+            var methodSymbol = (IMethodSymbol)c.Symbol;
 
-            if (methodSymbol != null &&
-                methodSymbol.IsExtern &&
+            if (methodSymbol.IsExtern &&
                 methodSymbol.IsPubliclyAccessible())
             {
-                c.ReportDiagnosticWhenActive(Diagnostic.Create(rule, methodDeclaration.Identifier.GetLocation(),
-                    MakeThisMethodPrivateMessage));
+                if (methodSymbol.DeclaringSyntaxReferences.FirstOrDefault()?.GetSyntax()
+                    is MethodDeclarationSyntax methodDeclaration)
+                {
+                    c.ReportDiagnosticWhenActive(Diagnostic.Create(rule, methodDeclaration.Identifier.GetLocation(),
+                        MakeThisMethodPrivateMessage));
+                }
             }
         }
 
         private static void ReportTrivialWrappers(SyntaxNodeAnalysisContext c)
         {
             var methodDeclaration = (MethodDeclarationSyntax)c.Node;
-            var methodSymbol = c.SemanticModel.GetDeclaredSymbol(methodDeclaration);

```

Comment:
Move symbol checks later in the code to ensure we do as many tests as possible on the syntax before getting the symbol.